### ISSUE TITLE
Compilation fails on rpi0

### ISSUE THREAD

**Issue by solsticedhiver** [falconindy/auracle#79]
I have 2 errors when compiling for rpi0 (raspberry pi 0 W, armv6h). Though I think, they are 2 differents issues, I, for now, open only one issue.

1. In an attempt to speed up complation, I have setup a *distcc* (with makepkg) with **cross-compilation** on rpi0 (distcc) and x86_64 (distccd), but the build fails with this error:

        FAILED: auracle 
        c++  -o auracle 'auracle@exe/src_auracle_main.cc.o' -Wl,--as-needed -Wl,--no-undefined -Wl,--gc-sections -march=armv6 -mfloat-abi=hard -mfpu=vfp -O2 -pipe -fstack-protector-strong -fno-plt -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Wl,--start-group libauracle.a libaur.a subprojects/abseil-cpp-20200225.2/libabsl_container_lib.a subprojects/abseil-cpp-20200225.2/libabsl_base_lib.a subprojects/abseil-cpp-20200225.2/libabsl_debugging_lib.a subprojects/abseil-cpp-20200225.2/libabsl_hash_lib.a subprojects/abseil-cpp-20200225.2/libabsl_synchronization_lib.a subprojects/abseil-cpp-20200225.2/libabsl_time_lib.a subprojects/abseil-cpp-20200225.2/libabsl_numeric_lib.a subprojects/abseil-cpp-20200225.2/libabsl_strings_lib.a subprojects/abseil-cpp-20200225.2/libabsl_status_lib.a subprojects/fmt-6.2.0/libfmt.a 
        /usr/lib/libalpm.so /usr/lib/libcurl.so /usr/lib/libsystemd.so -Wl,--end-group -pthread '-Wl,-rpath,$ORIGIN/:$ORIGIN/subprojects/abseil-cpp-20200225.2:$ORIGIN/subprojects/fmt-6.2.0:/usr/lib' -Wl,-rpath-link,/home/cassini/asp/auracle-git/src/auracle/build/ -Wl,-rpath-link,/home/cassini/asp/auracle-git/src/auracle/build/subprojects/abseil-cpp-20200225.2 -Wl,-rpath-link,/usr/lib -Wl,-rpath-link,/home/cassini/asp/auracle-git/src/auracle/build/subprojects/fmt-6.2.0
        /usr/bin/ld : subprojects/abseil-cpp-20200225.2/libabsl_base_lib.a(absl_base_internal_spinlock_wait.cc.o) : dans la fonction « absl::lts_2020_02_25::base_internal::SpinLockSuggestedDelayNS(int) » :
        spinlock_wait.cc:(.text+0x50) : référence indéfinie vers « __atomic_load_8 »
        /usr/bin/ld : spinlock_wait.cc:(.text+0x84) : référence indéfinie vers « __atomic_store_8 »
        collect2: erreur: ld a retourné le statut de sortie 1
        distcc[25599] ERROR: compile (null) on localhost failed
        [142/153] Compiling C++ object 'dependency_kind_test@exe/src_auracle_dependency_kind_test.cc.o'

rpi0 on archlinuxarm still uses gcc 9.3 and distccd cross-toolchain too if that matters

2. Then, when trying to compile auracle natively (without distcc): I got another error:

        [108/153] Compiling C++ object 'subprojects/abseil-cpp-20200225.2/ef33c1a@@absl_flags_lib@sta/absl_flags_internal_usage.cc.o'
        FAILED: subprojects/abseil-cpp-20200225.2/ef33c1a@@absl_flags_lib@sta/absl_flags_internal_usage.cc.o 
        c++ -Isubprojects/abseil-cpp-20200225.2/ef33c1a@@absl_flags_lib@sta -Isubprojects/abseil-cpp-20200225.2 -I../subprojects/abseil-cpp-20200225.2 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++17 -Wno-sign-compare -march=armv6 -mfloat-abi=hard -mfpu=vfp -O2 -fstack-protector-strong -fno-plt -D_FORTIFY_SOURCE=2 -fPIC -MD -MQ 'subprojects/abseil-cpp-20200225.2/ef33c1a@@absl_flags_lib@sta/absl_flags_internal_usage.cc.o' -MF 'subprojects/abseil-cpp-20200225.2/ef33c1a@@absl_flags_lib@sta/absl_flags_internal_usage.cc.o.d' -o 'subprojects/abseil-cpp-20200225.2/ef33c1a@@absl_flags_lib@sta/absl_flags_internal_usage.cc.o' -c ../subprojects/abseil-cpp-20200225.2/absl/flags/internal/usage.cc
        {standard input}: Assembler messages:
        {standard input}:2196: Warning: end of file not at end of a line; newline inserted
        {standard input}:3155: Error: bad arguments to instruction -- `ldr r3'
        c++: fatal error: Killed signal terminated program cc1plus
        compilation terminated.
        [109/153] Compiling C++ object 'subprojects/abseil-cpp-20200225.2/ef33c1a@@absl_flags_lib@sta/absl_flags_marshalling.cc.o'
        ninja: build stopped: subcommand failed.
        ==> ERROR: A failure occurred in build().
            Aborting...

I would understand that you would say that the first error is none of your concern

**Comment by falconindy** [comment]
Might be a problem with your toolchain if it's producing crap output for the assembler. Seems like something is truncated based on the warning. Perhaps you've run out of disk space somewhere?

**Comment by inglor** [comment]
I think it's coming from compiling the `abseil-cpp` and failing with memory issue or something. Check the AUR package [1] and see if you can compile it yourself (without distcc and with `-j1`). I'm trying now on my ARM but takes ages :)

Hopefully if you have it installed `meson` will pick it up and won't try to recompile it.

[1]: https://aur.archlinux.org/packages/abseil-cpp/

**Comment by falconindy** [comment]
Good catch. "Killed signal terminated program cc1plus" suggests that the compiler was sent a SIGKILL, probably due to OOM. You'd see evidence of that in dmesg.

**Comment by inglor** [comment]
OK with 4G swap (and 2G RAM) I managed to compile `abseil-cpp` but alas meson didn't detect it and proceeds to recompile it. I think maybe you can put a swap and set the `-j1` for meson and would probably compile - won't test it as `abseil-cpp` took a solid 2+ hours on my poor Odroid C2.

**Comment by solsticedhiver** [comment]
Ok. Yes I have OOM in the logs. so 2. is a no issue.

So it seems unreasonnable to attempt to compile it on the rpi0 after all  ...

Then, I come back to attempt of doing it with distcc and cross compile it. I'll talk to the distccd-armv6h packager to see if he knows what is going on.

I think you can close the issue, no ?

**Comment by falconindy** [comment]
It's not my code using atomics, it's abseil. I don't know much about ARM toolchains. This is why I don't mark arm as supposed in auracle's arch=() array on the AUR.

**Comment by inglor** [comment]
@falconindy is there a way in meson to instruct to use the `abseil` package is it's already installed (like `nlohmann_json` and only download it if it's not there.

**Comment by falconindy** [comment]
Not in the current build setup for auracle. abseil isn't really something that's meant to be built and installed. I definitely don't want to be linking against shared libs because there's zero guarantee of ABI stability.

**Comment by inglor** [comment]
@falconindy OK, understandable. @solsticedhiver The solution I found was make a swap (4G) set `MAKEFLAGS` `-j1` and use `clang` (version 10.0.0-3) to compile instead of `gcc` by modifying `PKGBUILD`. 
```
@@ -31,14 +31,14 @@ build() {

   [[ -d build ]] && meson_args+=(--wipe)

-  meson build "${meson_args[@]}"
+  CC=clang CXX=clang++ meson build "${meson_args[@]}"
```

**Comment by solsticedhiver** [comment]
@inglor one way to do it is with **qemu-user-static** and a *chroot* (I am using systemd-nspawn in fact, but arch-nspawn is good too) on **x86_64** where you can expect to have more RAM, disk space and CPU power.

https://wiki.archlinux.org/index.php/QEMU#Chrooting_into_arm/arm64_environment_from_x86_64

But this is quite complicated to do it because git segfaults inside the arm chroot. From what I have gathered, it might be because git use multiple threads which is not possible inside systemd-nspawn (?) Trying to recompile git without threads leads to nothing better. I don't know if the **NO_PTHREADS git build** is broken (again ?) or I did something wrong. Because even with the new git package, **git still segfaults**.

TLDR;
I got it working by doing this:
1. Setup the chroot (download the archlinuxarm tar ball, and unpack it)
2. Go to a directory inside the extracted home.
3. Download a snapshot of auracle-git package from aur with wget
4. Run `makepkg -o` to make makepkg fetch with git (outside the chroot !) and extract the source
5. chroot into the build, I use `systemd-nspawn -M rpi0 --bind /etc/resolv.conf -D $chroot_dir`
6. Install clang, and modify PKGBUILD to use clang
7. fix the path inside the git repo to get correct version for the package: change path from $chroot_dir/<path_inside_chroot> to <path_inside_chroot> in every file. Search for $chroot_dir with grep -r
7. Add `/usr/bin/core_perl` to `$PATH`: `PATH=$PATH:/usr/bin/core_perl`
7. run `makepkg -A -e` in the directory where auracle-git source are


P.S.: oh and you need to use **fakeroot-tcp** instead of the standard fakeroot package, because it fails too inside the chroot ....

**Comment by solsticedhiver** [comment]
My PR https://github.com/mesonbuild/abseil-cpp/pull/4 has been merged upstream; so this is a no issue now.

@falconindy please update the wrap file for abseil-cpp. there is a new release. thank you.
